### ISSUE TITLE
GIL Bound Metric

### ISSUE THREAD

**Issue by mrocklin** [dask/distributed#7290]
I'm chatting with @jerry-kobold and he's expressing curiousity about how to tell if his code is releasing the GIL or not.  I pointed him to https://github.com/jcrist/ptime , which makes it easy to test if an isolated function call releases the GIL.  I'm curious if we can also do this after the fact though.

I suspect that if I asked @gjoseph92 he would say "run pyspy and take look for some lock acquire call". This would be great, but it also requires some foresight / knowledge.

I'm curious if, instead, it's easy to periodically poll and build a metric for how bound the GIL is.  I imagine that if we were to poll every 100ms or so that we could get a metric that said something like "Some thread is waiting on the GIL 70% of the time, consider changing things around" or "Some thread is waiting on the GIL only 2% of the time, you're probably fine".

**Comment by gjoseph92** [comment]
py-spy `top` mode has a % GIL metric that's probably what you want: https://github.com/benfred/py-spy#top.

![](https://github.com/benfred/py-spy/raw/master/images/console_viewer.gif)

How you get that into a metric is a different question though (cc @benfred).

> I'm curious if, instead, it's easy to periodically poll and build a metric for how bound the GIL is

My intuition is that this would be impossible to do in pure Python. Obviously, when the GIL is held, you can't run anything in Python to check if it's held.

That's why any of these GIL-profiling tools (py-spy, [giltracer](https://github.com/maartenbreddels/per4m), [gil_load](https://github.com/chrisjbillington/gil_load)) operate outside of Python and poke in at it to sample whether the GIL is held or not.

I think the best you could do pure-Python is write a (pretty hot) loop in a thread that tries to do something which would be slowed down if there was GIL contention, and measure how much it's slowed down. Of course, by doing that, you'd be intentionally creating GIL contention. And what slows you down might not even be the GIL.

Here's a quick script that tries to do this. On my machine, it reports the low-contention case as 16% (py-spy reports ~2%), and full-contention as 70% (py-spy reports 100%). I probably got something wrong here so maybe you could make it more accurate, but not too promising initially.

<details>

```python
import collections
import time
import threading
import tqdm


delays = collections.deque(maxlen=10_000)


def sampler(stop: threading.Event, rate: int = 100) -> None:
    last = time.perf_counter_ns()
    interval = 1 / rate
    interval_ns = int(interval * 1e9)
    while not stop.is_set():
        time.sleep(interval)
        delay = (time.perf_counter_ns() - last) / interval_ns
        delays.append(delay)
        last = time.perf_counter_ns()


stop = threading.Event()
t = threading.Thread(target=sampler, args=(stop,))
t.start()

# Burn CPU; high GIL contention
for _ in tqdm.trange(10_000_000):
    pass

# Not much GIL contention
# for _ in tqdm.trange(50):
#     time.sleep(0.1)

stop.set()
t.join()

with open("delays.csv", "w", buffering=8192) as f:
    for d in delays:
        f.write(f"{d},\n")

# Print average
print(sum(delays) / len(delays) - 1)
```

</details>

---

I think the lightest-weight way to do this accurately (as in, something you'd be okay with having always on, minimal performance impact) would be a C extension which creates a C thread to periodically check if the GIL is held, then occasionally takes the GIL to report this back to Python at some interval.

**Comment by mrocklin** [comment]
I wonder if we could nerd snipe @saulshanabrook .  This seems like the kind of problem that he might enjoy.

**Comment by mrocklin** [comment]
Chatting with @thomasjpfan as well and he was curious about this topic

**Comment by mrocklin** [comment]
And apparently @mdroettboom

**Comment by mrocklin** [comment]
Thinking again about the API, maybe it makes sense to specify a time scale.  "I care about the GIL over the last five seconds".  Then maybe this is easy to implement as an exponentially weighted moving average (just one bit of state).  

```python
import gil_tracker

gil_tracker.start(timescale=datetime.timedelta(seconds=5))

gil_tracker.report()
5.3
```

**Comment by saulshanabrook** [comment]
I think you mean @mdboom

**Comment by thomasjpfan** [comment]
I ran into trouble trying to get the GIL state from `PyInterpreterState` because it is an opaque struct. This GIL state is shared across all threads, so tracking that would have been sufficient to monitor the GIL. I likely need to do something more clever to access the GIL like `py-spy` [1] or rethink my approach all together.


<details><summary><strong>[1]:</strong> How py-spy locates the GIL</summary>

`py-spy` builds Python with the private headers on and figures out the offsets for members in the opaque structs and stores the offset in [src/python_binds/mod.rs](https://github.com/benfred/py-spy/blob/master/src/python_bindings/mod.rs). The offsets can differ between patch versions of Python and platforms.

</details>

<details><summary><strong>[2]:</strong> Can we use PyGILState_Check?</summary>

`PyGILState_Check` only checks if the **current** thread holds the GIL, so it is not enough for this issue. The current thread is monitoring the GIL without holding the GIL itself.

</details>

**Comment by milesgranger** [comment]
Hi all,

@mrocklin pointed me in this direction, and finally had some time over the weekend to have a go at it. 
I also experienced the same limitations as explained by @thomasjpfan and others. So in my implementation [1], I used a separate thread to periodically re-acquire the GIL and measure the time taken and this over the runtime gives a metric which seems to be roughly on par with `py-spy` [2]. The use is similar to what was initially described here [3].

[1] [gilknocker](https://github.com/milesgranger/gilknocker) (forgive my lack of creativity in naming. :) )
[2] [py-spy top -- python tests/test_knockknock.py](https://github.com/milesgranger/gilknocker/blob/01e3a0914a3df3bacefb35a493f762a6d9a0caab/tests/test_knockknock.py#L76)
[3] [example use](https://github.com/milesgranger/gilknocker#use)

**Comment by mrocklin** [comment]
Super interesting @milesgranger .  Thanks!  

Some questions:
- Is this something that we could leave running constantly with mild (<1%) effect on performance?
- You mention the possibility of deadlocking.  How serious is this?
- Ideally we would be able to poll this periodically (every second or so).  I notice in your API it has more of a start/stop API.  Thoughts on if this could be modified to provide more constant feedback?

For context, I'd like to find something that we could install by default and just run all the time.  This would then be included as a plot in the Dask dashboard, and also as a Prometheus metric for general checking of all Dask clusters.  This would make it easy for us to then do post-hoc analysis on many historical clusters in an organization and make recommendations about things like worker size, libraries used, etc..

Thoughts?

**Comment by milesgranger** [comment]
> Is this something that we could leave running constantly with mild (<1%) effect on performance?

Good question. I suspect there is a balance between the configured monitoring interval and how contentious the GIL code being monitored is, changing how much effect his has on the program performance. I'll have to get back to you on this after setting up some benchmarks as time permits.

> You mention the possibility of deadlocking. How serious is this?

I'm quite certain this is caused by something from running the tests thru pytest. If instead I run the tests in an endless repetitive loop, as a script, I haven't been able to get it to deadlock. However, I don't know why it does so with pytest ~20-30% of the time, and whether or not the same underlying cause would exist in another environment. 

> Ideally we would be able to poll this periodically (every second or so). I notice in your API it has more of a start/stop API. Thoughts on if this could be modified to provide more constant feedback?

Think this is basically done in https://github.com/milesgranger/gilknocker/pull/2
Pretty easy adjustment, but `stop` ought to be called at some point to close the thread, otherwise I suppose it'll just run until the parent process exits. Anyhow, the contention metric will be continuously updated so can read it anytime you like.

**Comment by mrocklin** [comment]
> I'll have to get back to you on this after setting up some benchmarks as time permits.

I'll take what I can get 🙂

> otherwise I suppose it'll just run until the parent process exits

This seems fine to me!

> Anyhow, the contention metric will be continuously updated so can read it anytime you like

Ideally we would have some level of timeliness here.  In the context of Dask I'd want to notice, for example, "Hrm, during this period where I'm reading Parquet I notice that the GIL is way more occupied while during this period when I'm doing a join it isn't occupied that much.  I know, maybe I'll change my parquet code to ..."

My thought on this before was to do an exponentially weighted moving average. Thoughts?  Doable?

**Comment by milesgranger** [comment]
Ah, then from the `gilknocker` lib perspective, adding a `reset` to the contention metric, which is otherwise updated in real-time, would also be quite easy. Then the user can implement their own logic for gauging specific chunks and/or doing their own weighted moving average by choosing when to reset the metric and when to read it.

**Comment by mrocklin** [comment]
So presumably with prometheus we would do something like the following ...

```python
from gilknocker import KnockKnock

knocker = KnockKnock(interval_micros=1000, timeout_secs=None)
knocker.start()

def report_prometheus_gil_metric():
    out = knocker.contention_metric
    knocker.reset()
    return out
```

**Comment by mrocklin** [comment]
@fjetter @ntabris any interest in this?  What would you need in order to consume this?

@milesgranger thoughts on publishing?  Perhaps conda-forge would be especially helpful given the Rust dependency?

**Comment by milesgranger** [comment]
All wheels for most platforms are [already on PyPi](https://pypi.org/project/gilknocker/#files), as they have no other Python dependencies, I'd like to avoid  [duplicate compilation work](https://github.com/milesgranger/gilknocker/blob/main/.github/workflows/CI.yml) where possible. Does conda have a way to just take the existing wheels from PyPI?

**Comment by mrocklin** [comment]
Hrm, I gave this a spin and had a non-positive time.  Suggestions?

```python
In [1]: import gilknocker

In [2]: gilknocker.KnockKnock?
Init signature: gilknocker.KnockKnock(self, /, *args, **kwargs)
Docstring:
Struct for polling, knocking on the GIL,
checking if it's locked in the current thread

Example
-------
    ```python
from gilknocker import KnockKnock
knocker = KnockKnock(100)  # try to reacquire the gil every 100 microseconds
knocker.start()
... some smart code ...
knocker.stop()
knocker.contention_metric  # float between 0-1 indicating GIL contention
    ```
Type:           type
Subclasses:

In [3]: knocker = gilknocker.KnockKnock(interval_micros=1000, timeout_secs=None)
   ...:

In [4]: knocker.start()

In [5]: knocker.contention_metric

In [6]: knocker.contention_metric

In [7]: knocker.stop()
---------------------------------------------------------------------------
TimeoutError                              Traceback (most recent call last)
Cell In [7], line 1
----> 1 knocker.stop()

TimeoutError: Failed to join knocker thread.

In [8]: knocker.contention_metric

In [9]: knocker = gilknocker.KnockKnock(interval_micros=1000, timeout_secs=5)

In [10]: knocker.start()

In [11]: knocker.contention_metric

In [12]: knocker.stop()
---------------------------------------------------------------------------
TimeoutError                              Traceback (most recent call last)
Cell In [12], line 1
----> 1 knocker.stop()

TimeoutError: Failed to join knocker thread.

In [14]: gilknocker.__version__
Out[14]: '0.1.0'
```

**Comment by milesgranger** [comment]
Version 0.1.0 on PyPI doesn't impl what we talked about, continuously updating the contention metric. So in that version contention metric won't have a value until stop, here's what I have:

```python
In [1]: from gilknocker import KnockKnock

In [2]: knocker = KnockKnock(1000, 1)

In [3]: knocker.start()

In [4]: knocker.stop()

In [5]: knocker.contention_metric
Out[5]: 0.007168566808104515
```

It's interesting though to note you get the TimeoutError, and I can't seem to get that.  Curious if you could run it from a file? (`python my_test_script.py`) and see if TimeoutError happens there.

I'll let you know when I finish up the PR and release 0.2.0; maybe we can continue the discussion on an issue [over on there](https://github.com/milesgranger/gilknocker)

**Comment by fjetter** [comment]
Tested briefly 0.2.0 and it looks promising. API wise this is sufficient to get started. 
We will likely want to calculate some moving averages over different intervals instead of having a single point estimator but we can do this in Python as well and iterate on the API as necessary.

Regarding overhead, I'm not super concerned about it but I guess this depends on the sampling rate and we'll need to run a couple of tests.

> So presumably with prometheus we would do something like the following ...

Considering that we'd want to build a bokeh dashboard with this, I guess we'll rather collect this info in a PeriodicCallback and cache it on the Server. [`SystemMonitor`](https://github.com/dask/distributed/blob/main/distributed/system_monitor.py) might be a good fit since it already implements a periodic update and is collecting things from psutil and nvml

**Comment by jjerphan** [comment]
Some distribution of CPython (e.g. conda-forge's) come with functions' symbols for interacting with the GIL (namely `take_gil`/`drop_gil`).

On Linux, calls and returns from those functions are traceable using [`perf(1)`](https://man7.org/linux/man-pages/man1/perf.1.html) by defining uprobes, provided that the `CAP_PERFMON` is set (more about this on the [`capabilities(7)` man page](
https://man7.org/linux/man-pages/man7/capabilities.7.html). Tracing this `perf(1)` generally comes with little impact on performance.

@mrocklin: to you, would having Linux with this capacity being activated in production be possible as a scenario?

**Comment by jjerphan** [comment]
Some distributions of CPython (e.g. conda-forge's) come with symbols of functions interacting with the GIL (namely `take_gil`/`drop_gil`).

On Linux, calls and returns from those functions are traceable using [`perf(1)`](https://man7.org/linux/man-pages/man1/perf.1.html) by defining uprobes, provided that the `CAP_PERFMON` capability is set (more about this on the [`capabilities(7)` man page](
https://man7.org/linux/man-pages/man7/capabilities.7.html)). Tracing using `perf(1)` generally comes with little impact on performance, yet it necessitates running it outside of CPython.

Would having Linux machines with this capability being activated in production be conceivable?

**Comment by gjoseph92** [comment]
To me, tracing using perf would be a good solution. Something self-contained and cross-platform is _better_ (especially long-term), but something that automated setting up and reporting on the uprobes would be fine initially (maybe building off of https://github.com/maartenbreddels/per4m?). Having `CAP_PERFMON` set in production also seems fine to me.

If we got something working with those caveats using perf, I'd be very happy. If it seemed super useful, then maybe we'd want to think about how to generalize it.

**Comment by mrocklin** [comment]
Is the solution linked to above *good enough*?  If so, I'm inclined to just go ahead.  If it's likely to cause issues (performance hit, more contention) then yeah, we should look at other approaches.

**Comment by milesgranger** [comment]
FWIW, I [added some benchmarks](https://github.com/milesgranger/gilknocker#how-will-this-impact-my-program), and it seems for GIL bound tasks, it has pretty much no effect (within the bounds of the control case, of no gil monitoring). Then for GIL light tasks, as suspected the program will be (mildly) impacted in correlation to the higher frequency monitoring.

**Comment by jjerphan** [comment]
> Is the solution linked to above good enough?

Are you referring to [`maartenbreddels/per4m`](https://github.com/maartenbreddels/per4m)?

**Comment by mrocklin** [comment]
I refer to https://github.com/milesgranger/gilknocker

**Comment by jjerphan** [comment]
I've been playing with [`milesgranger/gilknocker`](https://github.com/milesgranger/gilknocker) a bit locally, and 0.2.0 seems to have fixes previously observed problems.

@milesgranger: this tool might be useful for some work in other libraries like scikit-learn or joblib. :+1: 

I think it makes sense to first test it on various scenario without dask, especially ones involving C extension with and without `pthread`- and `OpenMP`-multi-threading, before trying to integrate it in the dashboard reports.

**Comment by gjoseph92** [comment]
@crusaderky doesn't that difference include waiting on locks/IO besides the GIL? For example, if a user task was blocked on a `read`, or on some internal mutex in a thread-unsafe library, wouldn't that also lead to a large gap between `time` and `thread_time`?

**Comment by crusaderky** [comment]
@gjoseph92 yes, you need to decorate all the code that is known to perform I/O or internal locking.
e.g. in order to produce https://gist.github.com/crusaderky/a97f870c51260e63a1c14c20b762f666 I had to patch dask/dask:
 (I now realised that the published notebook wiped out the snippet for some reason)

```python
#%% raw
--- a/dask/dataframe/io/parquet/core.py
+++ b/dask/dataframe/io/parquet/core.py
@@ -27,6 +27,8 @@ from dask.highlevelgraph import HighLevelGraph
 from dask.layers import DataFrameIOLayer
 from dask.utils import apply, import_required, natural_sort_key, parse_bytes
 
+from distributed.metrics import context_meter
+
 __all__ = ("read_parquet", "to_parquet")
 
 NONE_LABEL = "__null_dask_index__"
@@ -158,6 +160,7 @@ class ToParquetFunctionWrapper:
             self.kwargs_pass,
         )
 
+    @context_meter.meter("I/O")
     def __call__(self, df, block_index: tuple[int]):
         # Get partition index from block index tuple
         part_i = block_index[0]
@@ -643,6 +646,7 @@ def check_multi_support(engine):
     return hasattr(engine, "multi_support") and engine.multi_support()
 
 
+@context_meter.meter("I/O")
 def read_parquet_part(
     fs, engine, meta, part, columns, index, use_nullable_dtypes, kwargs
 ):
```

Alternatively to the above, you can just break down the 'thread-noncpu' by task prefix, look at the big offenders, and manually identify those that are known to be I/O related - a fairly straightforward operation in most cases.

**Comment by milesgranger** [comment]
Closing as I believe this has been addressed in [2023.4.1](https://blog.coiled.io/blog/gil-monitoring.html)
